### PR TITLE
ociruntime: replace tar extraction subprocess with in-process

### PR DESCRIPTION
**Pull request by sluongng**
Address the TODO comment.

### CODE REVIEW

**Review by bduffany — COMMENTED** [review]
I was holding off on this TODO because I wanted to make sure the performance was on-par. Any chance you could run some quick tests with larger images with several layers (on a fast network connection, clearing cache between runs) to make sure it's not significantly slower and CPU usage is not significantly higher?

**Comment by sluongng** [comment]
I applied this patch on top of master

```
diff --git a/enterprise/server/remote_execution/containers/ociruntime/BUILD b/enterprise/server/remote_execution/containers/ociruntime/BUILD
index 5d70c3ab1d..4c3b386f12 100644
--- a/enterprise/server/remote_execution/containers/ociruntime/BUILD
+++ b/enterprise/server/remote_execution/containers/ociruntime/BUILD
@@ -48,6 +48,7 @@ go_test(
         "test.workload-isolation-type": "firecracker",
         "test.container-image": "docker://gcr.io/flame-public/net-tools@sha256:ac701954d2c522d0d2b5296323127cacaaf77627e69db848a8d6ecb53149d344",
         "test.EstimatedComputeUnits": "2",
+        "test.EstimatedFreeDiskBytes": "4.5GB",
     },
     # NOTE: if testing locally, use --test_sharding_strategy=disabled to work
     # around the networking package not supporting cross-process locks.
diff --git a/enterprise/server/remote_execution/containers/ociruntime/ociruntime_test.go b/enterprise/server/remote_execution/containers/ociruntime/ociruntime_test.go
index 4d307ad96d..278dd16af3 100644
--- a/enterprise/server/remote_execution/containers/ociruntime/ociruntime_test.go
+++ b/enterprise/server/remote_execution/containers/ociruntime/ociruntime_test.go
@@ -1119,3 +1119,33 @@ func hasMountPermissions(t *testing.T) bool {
        require.NoError(t, err, "unmount")
        return true
 }
+
+func TestPullImage(t *testing.T) {
+       for _, tc := range []struct {
+               name  string
+               image string
+       }{
+               {
+                       name:  "dockerhub_busybox",
+                       image: "busybox:latest",
+               },
+               {
+                       name:  "ghcr_nix",
+                       image: "ghcr.io/avdv/nix-build@sha256:5f731adacf7290352fed6c1960dfb56ec3fdb31a376d0f2170961fbc96944d50",
+               },
+               {
+                       name:  "executor_image",
+                       image: "gcr.io/flame-public/buildbuddy-executor-enterprise:latest",
+               },
+       } {
+               t.Run(tc.name, func(t *testing.T) {
+                       layerDir := t.TempDir()
+                       imgStore := ociruntime.NewImageStore(layerDir)
+
+                       ctx := context.Background()
+                       img, err := imgStore.Pull(ctx, tc.image, oci.Credentials{})
+                       require.NoError(t, err)
+                       require.NotNil(t, img)
+               })
+       }
+}
```

Then ran it with 
```
bazel test --config=remote --test_tag_filters=+docker enterprise/server/remote_execution/containers/ociruntime/... --test_filter=TestPullImage/executor_image --runs_per_test=10 --test_sharding_strategy=disabled
```

Repeated the same test with this PR on top.

- Before: https://buildbuddy.buildbuddy.io/invocation/4caca435-a067-470b-b414-ca88ef686951?executionFilter=test#execution

- After: https://buildbuddy.buildbuddy.io/invocation/41f236d6-e5bb-4945-b777-4c059ec7c33e?executionFilter=test#execution

I think the Go approach could be slightly slower and higher memory vs pipe to `tar`. With this small sample size the Executions, I think they are negligible. I suspect if the Go app is saturated, gc will kick in a bit more eagerly though.

**Review by bduffany — APPROVED** [review]

**Review by bduffany — APPROVED** [review]
LGTM assuming the `TestFileOwnership` test passes after you rebase (we aren't passing `--no-same-owner` anymore)

File: enterprise/server/remote_execution/containers/ociruntime/ociruntime.go
Comment (sluongng) [reply enterprise/server/remote_execution/containers/ociruntime/ociruntime.go:1267]: That's what I expect `filepath.Clean()` to handle https://cs.opensource.google/go/go/+/refs/tags/go1.23.3:src/path/filepath/path_test.go;l=54-69;drc=1eee17dc4e796c2fe60e8f9e82b66c00610adc31.

So combining clean + trim prefix + join should get rid of most cases(?) https://go.dev/play/p/6o4loXyQdNv

Do you think we should err out instead? Or we should add a `strings.Contains` check + err after this as a precaution?

Comment (fmeum) [reply enterprise/server/remote_execution/containers/ociruntime/ociruntime.go:1267]: `Clean` wouldn't help in case `tempUnpackDir` is an absolute path, tar entries could still navigate anywhere on the file system. But even if we check for `..`, this is still possible by creating a symlink first and then using it to jump. 

I think that we should first clarify what the implications of that happening are going to be: is this flexibility a feature, a foot gun or may it have security implications?

Comment (sluongng) [reply enterprise/server/remote_execution/containers/ociruntime/ociruntime.go]: Playing with the playground a bit more, because we are using `filepath.Join()`, we should not need to apply `TrimPrefix` to the cleaned path to remove leading `/`.

So I removed TrimPrefix as well.

Comment (bduffany) [reply enterprise/server/remote_execution/containers/ociruntime/ociruntime.go]: > Do you think we should err out instead? 

@sluongng yes, I think we should return an error. If a tar file contains a file name like `../foo` we only have a couple of options: (a) ignore it, (b) return an error, (c) do something incorrect or insecure. I think (b) is the best option because the "ignore" option is sort of failing "silently" which might be confusing imo

Comment (sluongng) [reply enterprise/server/remote_execution/containers/ociruntime/ociruntime.go]: I added the err to the general case as well as the hardlink case.

I checked for both `strings.Contain` as well as `filepath.Clean` diff so that we can avoid cases like `dir/some..file/`.

Comment (fmeum) [reply enterprise/server/remote_execution/containers/ociruntime/ociruntime.go]: Isn't `dir/some..file/` totally fine? Only `../`, `/../` and `/..` could be problematic.

Comment (fmeum) [reply enterprise/server/remote_execution/containers/ociruntime/ociruntime.go]: `/` is also a path separator on Windows. While that's not our focus here, could you convert this with `filepath.ToSlash` and then split on `/`?

File: enterprise/server/remote_execution/containers/ociruntime/ociruntime.go
Comment (bduffany) [inline enterprise/server/remote_execution/containers/ociruntime/ociruntime.go:1316]: The `cleanHeaderName != header.Name` check seems undesired - `filepath.Clean("../foo")` returns `"../foo"` so the check would not be run.

```suggestion
		if slices.Contains(filepath.Split(header.Name), "..") {
			return status.UnavailableError("tar entry contains disallowed parent directory reference", header.Name)
		}
```

Comment (sluongng) [reply enterprise/server/remote_execution/containers/ociruntime/ociruntime.go:1316]: Oh wat. We probably dont want filepath.Split, but i agree with the fix

Comment (bduffany) [reply enterprise/server/remote_execution/containers/ociruntime/ociruntime.go]: woops, filepath.Split doesn't do what I thought it did :skull:

**Comment by sluongng** [comment]
I learned about https://github.com/google/safeopen/blob/main/safeopen_linux.go over the weekend as well as https://github.com/golang/go/issues/67002#issuecomment-2460584052 (coming to Go 1.24). But I think these could be added later in a different PR?

Thoughts on merging the current PR as-is @fmeum @bduffany ?

**Comment by sluongng** [comment]
Also learned about https://pkg.go.dev/github.com/landlock-lsm/go-landlock/landlock today which could be useful if we want to do these in a subprocess in the future.

**Pull request merged**
